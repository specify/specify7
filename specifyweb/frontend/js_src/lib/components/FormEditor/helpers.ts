import type { RR } from '../../utils/types';
import type { Tables } from '../DataModel/types';

export const tablesWithBusRulesIn6 = new Set([
  'AccessionAgentBusRules',
  'AccessionAuthorizationBusRules',
  'AccessionBusRules',
  'AddressBusRules',
  'AgentBusRules',
  'AppraisalBusRules',
  'AttachmentBusRules',
  'AuthorBusRules',
  'BioStratBusRules',
  'BorrowBusRules',
  'CatalogNumberingSchemeBusRules',
  'CatAutoNumberingSchemeBusRules',
  'CollectingEventBusRules',
  'CollectingTripAuthorizationBusRules',
  'CollectingTripBusRules',
  'CollectionBusRules',
  'CollectionObjectBusRules',
  'CollectorBusRules',
  'ConservDescriptionBusRules',
  'ConservEventBusRules',
  'ContainerBusRules',
  'DataTypeBusRules',
  'DeaccessionAgentBusRules',
  'DeterminationBusRules',
  'DeterminerBusRules',
  'DisciplineBusRules',
  'DivisionBusRules',
  'DNASequenceBusRules',
  'DNASequencingRunBusRules',
  'ExchangeOutBusRules',
  'ExchangeOutPrepBusRules',
  'ExtractorBusRules',
  'FieldNotebookBusRules',
  'FieldNotebookPageSetBusRules',
  'FundingAgentBusRules',
  'GeographyBusRules',
  'GeologicTimePeriodBusRules',
  'GiftBusRules',
  'GiftPreparationBusRules',
  'GroupPersonBusRules',
  'InfoRequestBusRules',
  'LithoStratBusRules',
  'LoanBusRules',
  'LoanGiftShipmentBusRules',
  'LoanPreparationBusRules',
  'LoanReturnPreparationBusRules',
  'LocalityBusRules',
  'PcrPersonBusRules',
  'PermitBusRules',
  'PickListBusRules',
  'PreparationBusRules',
  'PrepTypeBusRules',
  'SpecifyUserBusRules',
  'StorageBusRules',
  'TaxonBusRules',
  'TreatmentEventBusRules',
  'TreeDefBusRules',
]);

/**
 * Most of the time business rules class name can be inferred from table name.
 * Exceptions:
 */
export const businessRulesOverride: Partial<RR<keyof Tables, string>> = {
  Shipment: 'LoanGiftShipment',
  CollectionObjectAttachment: 'Attachment',
};

export const getBusinessRuleClassFromTable = (
  tableName: keyof Tables
): string =>
  businessRulesOverride[tableName] === undefined
    ? tablesWithBusRulesIn6.has(`${tableName}BusRules`)
      ? `edu.ku.brc.specify.datamodel.busrules.${tableName}BusRules`
      : ''
    : `edu.ku.brc.specify.datamodel.busrules.${businessRulesOverride[
        tableName
      ]!}BusRules`;
