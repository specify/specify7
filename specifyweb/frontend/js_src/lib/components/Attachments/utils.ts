import { f } from '../../utils/functools';
import { filterArray } from '../../utils/types';
import type { Relationship } from '../DataModel/specifyField';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { genericTables, getTable } from '../DataModel/tables';
import { hasTablePermission } from '../Permissions/helpers';

export const attachmentRelatedTables = f.store(() =>
  Object.keys(genericTables).filter((tableName) =>
    tableName.endsWith('Attachment')
  )
);

export const allTablesWithAttachments = f.store(() =>
  filterArray(
    attachmentRelatedTables().map((tableName) =>
      getTable(tableName.slice(0, -1 * 'Attachment'.length))
    )
  )
);

/** Exclude tables without read access */
export const tablesWithAttachments = f.store(() =>
  allTablesWithAttachments().filter((model) =>
    hasTablePermission(model.name, 'read')
  )
);

/**
 * Gets the relationship to a Table's attachment table.
 * e.g., the collectionObjectAttachments relationship from CollectionObject to CollectionObjectAttachments
 * In most cases the relationship is called tableNameAttachments, although fetching the relationship
 * following that schema will not always be consitent as the following models relationship to their
 * attachment table is simply called `attachments`:
 * - DNASequence
 * - FieldNotebook
 * - FieldNotebookPage
 * - FieldNotebookPageSet
 */
export const getAttachmentRelationship = (
  table: SpecifyTable
): Relationship | undefined => {
  if (table.name === 'Attachment') return undefined;
  const commonRelationship = table.field[
    `${table.name[0].toLowerCase()}${table.name.slice(1)}Attachments`
  ] as Relationship | undefined;

  return commonRelationship === undefined
    ? table.relationships.find((relationship) => {
        const relatedModel = relationship.relatedTable.name;
        return (
          relatedModel !== 'Attachment' && relatedModel.endsWith('Attachment')
        );
      })
    : commonRelationship;
};
