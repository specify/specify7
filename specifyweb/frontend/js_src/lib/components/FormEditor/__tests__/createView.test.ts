import { requireContext } from '../../../tests/helpers';
import type { IR } from '../../../utils/types';
import { ensure, localized } from '../../../utils/types';
import { tables } from '../../DataModel/tables';
import type { ViewDefinition } from '../../FormParse';
import { formatXmlForTests } from '../../Syncer/__tests__/utils';
import { jsonToXml } from '../../Syncer/xmlToJson';
import { xmlToString } from '../../Syncer/xmlToString';
import { createViewDefinition, exportsForTests } from '../createView';
import type { ViewSets } from '../spec';

requireContext();

const { tablesWithFormTable } = exportsForTests;

test('Tables with form tables computed correctly', () =>
  expect(tablesWithFormTable()).toMatchSnapshot());

const viewDefinition: ViewDefinition = {
  name: localized('CollectionObjectAttachment'),
  class: 'edu.ku.brc.specify.datamodel.CollectionObjectAttachment',
  busrules: 'edu.ku.brc.specify.datamodel.busrules.AttachmentBusRules',
  resourcelabels: 'false',
  altviews: {},
  viewdefs: {
    'CollectionObjectAttachment IconView': `<viewdef
  type="iconview"
  name="CollectionObjectAttachment IconView"
  class="edu.ku.brc.specify.datamodel.CollectionObjectAttachment"
  gettable="edu.ku.brc.af.ui.forms.DataGetterForObj"
  settable="edu.ku.brc.af.ui.forms.DataSetterForObj">
  <desc><![CDATA[The ObjectAttachment Icon Viewer]]></desc>
</viewdef>`,
    'CollectionObjectAttachment Table': `<viewdef
  type="formtable"
  name="CollectionObjectAttachment Table"
  class="edu.ku.brc.specify.datamodel.CollectionObjectAttachment"
  gettable="edu.ku.brc.af.ui.forms.DataGetterForObj"
  settable="edu.ku.brc.af.ui.forms.DataSetterForObj">
  <desc><![CDATA[ObjectAttachment grid view.]]></desc>
  <definition>ObjectAttachment Form</definition>
</viewdef>`,
    'CollectionObjectAttachment Form': `<viewdef
  type="form"
  name="CollectionObjectAttachment Form"
  class="edu.ku.brc.specify.datamodel.CollectionObjectAttachment"
  gettable="edu.ku.brc.af.ui.forms.DataGetterForObj"
  settable="edu.ku.brc.af.ui.forms.DataSetterForObj"
  useresourcelabels="true">
  <desc><![CDATA[The CollectionObjectAttachment form.]]></desc>
  <!-- <columnDef>110px,2dlu,p:g,5dlu,100px,2dlu,85px</columnDef> -->
  <columnDef>p,5dlu,p:g</columnDef>
  <rowDef auto="true" cell="p" sep="2px"/>
  <rows />
</viewdef>`,
    'ObjectAttachment Form': `<viewdef type="form" name="ObjectAttachment Form" class="edu.ku.brc.specify.datamodel.ObjectAttachmentIFace" gettable="edu.ku.brc.af.ui.forms.DataGetterForObj" settable="edu.ku.brc.af.ui.forms.DataSetterForObj" useresourcelabels="true">
  <desc>The ObjectAttachment form.</desc>
  <columnDef>p,2px,p:g</columnDef>
  <rowDef auto="true" cell="p" sep="2px" />
  <rows />
</viewdef>`,
  },
  view: `<view name="CollectionObjectAttachment" class="edu.ku.brc.specify.datamodel.CollectionObjectAttachment" busrules="edu.ku.brc.specify.datamodel.busrules.AttachmentBusRules" resourcelabels="false">
  <desc>The Collection Object-Attachment View.</desc>
  <altviews>
    <altview name="CollectionObjectAttachment Icon View" viewdef="CollectionObjectAttachment IconView" mode="view" />
    <altview name="CollectionObjectAttachment Icon Edit" viewdef="CollectionObjectAttachment IconView" mode="edit" />
    <altview name="CollectionObjectAttachment Table View" viewdef="CollectionObjectAttachment Table" mode="view" />
    <altview name="CollectionObjectAttachment Table Edit" viewdef="CollectionObjectAttachment Table" mode="edit" />
    <altview name="CollectionObjectAttachment Form View" viewdef="CollectionObjectAttachment Form" label="Form" mode="view" default="true" />
    <altview name="CollectionObjectAttachment Form Edit" viewdef="CollectionObjectAttachment Form" label="Form" mode="edit" />
  </altviews>
</view>`,
  viewsetName: 'Global',
  viewsetLevel: 'Backstop',
  viewsetSource: 'disk',
  viewsetId: null,
  viewsetFile: 'backstop/global.views.xml',
};

const viewSets = (): ViewSets =>
  ensure<ViewSets>()({
    views: [
      {
        name: localized('CollectionObjectAttachment'),
        description: 'The Collection Object-Attachment View.',
        businessRules: localized(
          'edu.ku.brc.specify.datamodel.busrules.AttachmentBusRules'
        ),
        legacyIsInternal: undefined,
        legacyIsExternal: undefined,
        legacyTable: undefined,
        title: localized(''),
        altViews: {
          legacyDefaultMode: undefined,
          legacySelector: undefined,
          altViews: [
            {
              name: 'CollectionObjectAttachment Icon View',
              viewDef: localized('CollectionObjectAttachment IconView'),
              mode: 'view',
              default: false,
              legacyTitle: undefined,
              legacyLabel: undefined,
              legacyValidated: undefined,
              legacySelectorValue: undefined,
            },
            {
              name: 'CollectionObjectAttachment Icon Edit',
              viewDef: localized('CollectionObjectAttachment IconView'),
              mode: 'edit',
              default: false,
              legacyTitle: undefined,
              legacyLabel: undefined,
              legacyValidated: undefined,
              legacySelectorValue: undefined,
            },
            {
              name: 'CollectionObjectAttachment Table View',
              viewDef: localized('CollectionObjectAttachment Table'),
              mode: 'view',
              default: false,
              legacyTitle: undefined,
              legacyLabel: undefined,
              legacyValidated: undefined,
              legacySelectorValue: undefined,
            },
            {
              name: 'CollectionObjectAttachment Table Edit',
              viewDef: localized('CollectionObjectAttachment Table'),
              mode: 'edit',
              default: false,
              legacyTitle: undefined,
              legacyLabel: undefined,
              legacyValidated: undefined,
              legacySelectorValue: undefined,
            },
            {
              name: 'CollectionObjectAttachment Form View',
              viewDef: localized('CollectionObjectAttachment Form'),
              mode: 'view',
              default: true,
              legacyTitle: undefined,
              legacyLabel: localized('Form'),
              legacyValidated: undefined,
              legacySelectorValue: undefined,
            },
            {
              name: 'CollectionObjectAttachment Form Edit',
              viewDef: localized('CollectionObjectAttachment Form'),
              mode: 'edit',
              default: false,
              legacyTitle: undefined,
              legacyLabel: localized('Form'),
              legacyValidated: undefined,
              legacySelectorValue: undefined,
            },
          ],
        },
        legacyUseBusinessRules: true,
        legacyResourceLabels: false,
        table: tables.CollectionObjectAttachment,
      },
    ],
    viewDefs: [
      {
        name: localized('CollectionObjectAttachment Table'),
        type: 'formtable',
        legacyGetTable: localized('edu.ku.brc.af.ui.forms.DataGetterForObj'),
        legacySetTable: localized('edu.ku.brc.af.ui.forms.DataSetterForObj'),
        legacyEditableDialog: true,
        legacyUseResourceLabels: undefined,
        raw: {
          type: 'XmlNode',
          tagName: 'viewdef',
          attributes: {},
          children: [],
        },
        table: tables.CollectionObjectAttachment,
        legacyTable: undefined,
      },
      {
        name: localized('CollectionObjectAttachment Form'),
        type: 'form',
        legacyGetTable: localized('edu.ku.brc.af.ui.forms.DataGetterForObj'),
        legacySetTable: localized('edu.ku.brc.af.ui.forms.DataSetterForObj'),
        legacyEditableDialog: true,
        legacyUseResourceLabels: true,
        raw: {
          type: 'XmlNode',
          tagName: 'viewdef',
          attributes: {},
          children: [],
        },
        table: tables.CollectionObjectAttachment,
        legacyTable: undefined,
      },
      {
        name: localized('CollectionObjectAttachment IconView'),
        type: 'iconview',
        legacyGetTable: localized('edu.ku.brc.af.ui.forms.DataGetterForObj'),
        legacySetTable: localized('edu.ku.brc.af.ui.forms.DataSetterForObj'),
        legacyEditableDialog: true,
        legacyUseResourceLabels: undefined,
        raw: {
          type: 'XmlNode',
          tagName: 'viewdef',
          attributes: {},
          children: [],
        },
        table: tables.CollectionObjectAttachment,
        legacyTable: undefined,
      },
      {
        name: localized('ObjectAttachment Form'),
        type: 'form',
        legacyGetTable: localized('edu.ku.brc.af.ui.forms.DataGetterForObj'),
        legacySetTable: localized('edu.ku.brc.af.ui.forms.DataSetterForObj'),
        legacyEditableDialog: true,
        legacyUseResourceLabels: true,
        raw: {
          type: 'XmlNode',
          tagName: 'viewdef',
          attributes: {},
          children: [],
        },
        table: undefined,
        legacyTable: 'edu.ku.brc.specify.datamodel.ObjectAttachmentIFace',
      },
    ],
    name: viewDefinition.name,
  });

/** Reformat the result to reduce snapshot size and make it more readable */
const processViewSet = (viewSet: ViewSets): IR<unknown> =>
  // Get rid of symbols
  JSON.parse(
    JSON.stringify({
      ...viewSet,
      viewDefs: viewSet.viewDefs.map(({ raw, ...rest }) => ({
        ...rest,
        raw: formatXmlForTests(
          xmlToString(jsonToXml({ ...raw, attributes: {} }), false)
        ),
      })),
    })
  );

test('Create new view definition', () =>
  expect(
    processViewSet(
      createViewDefinition(
        viewSets(),
        localized('CollectionObjectAttachment_2'),
        tables.CollectionObjectAttachment,
        'new'
      )
    )
  ).toEqual({
    viewDefs: [
      {
        legacyEditableDialog: true,
        legacyGetTable: 'edu.ku.brc.af.ui.forms.DataGetterForObj',
        legacySetTable: 'edu.ku.brc.af.ui.forms.DataSetterForObj',
        name: 'CollectionObjectAttachment Table',
        raw: '<viewdef/>',
        table: '[table CollectionObjectAttachment]',
        type: 'formtable',
      },
      {
        legacyEditableDialog: true,
        legacyGetTable: 'edu.ku.brc.af.ui.forms.DataGetterForObj',
        legacySetTable: 'edu.ku.brc.af.ui.forms.DataSetterForObj',
        legacyUseResourceLabels: true,
        name: 'CollectionObjectAttachment Form',
        raw: '<viewdef/>',
        table: '[table CollectionObjectAttachment]',
        type: 'form',
      },
      {
        legacyEditableDialog: true,
        legacyGetTable: 'edu.ku.brc.af.ui.forms.DataGetterForObj',
        legacySetTable: 'edu.ku.brc.af.ui.forms.DataSetterForObj',
        name: 'CollectionObjectAttachment IconView',
        raw: '<viewdef/>',
        table: '[table CollectionObjectAttachment]',
        type: 'iconview',
      },
      {
        legacyEditableDialog: true,
        legacyGetTable: 'edu.ku.brc.af.ui.forms.DataGetterForObj',
        legacySetTable: 'edu.ku.brc.af.ui.forms.DataSetterForObj',
        legacyTable: 'edu.ku.brc.specify.datamodel.ObjectAttachmentIFace',
        legacyUseResourceLabels: true,
        name: 'ObjectAttachment Form',
        raw: '<viewdef/>',
        type: 'form',
      },
      {
        legacyEditableDialog: true,
        name: 'CollectionObjectAttachment_2',
        raw: "<viewdef>\n  <desc>The CollectionObjectAttachment Table</desc>\n  <enableRules/>\n  <columnDef>p,2px,p,2px,p,2px,p,2px,p,2px,p,p:g</columnDef>\n  <rowDef auto='true' cell='p' sep='2px'/>\n  <rows>\n    <row/>\n  </rows>\n</viewdef>",
        table: '[table CollectionObjectAttachment]',
        type: 'form',
      },
      {
        legacyEditableDialog: true,
        name: 'CollectionObjectAttachment_2 Icon',
        raw: '<viewdef>\n  <desc>The Attachments Icon Viewer</desc>\n</viewdef>',
        table: '[table CollectionObjectAttachment]',
        type: 'iconview',
      },
      {
        legacyEditableDialog: true,
        name: 'CollectionObjectAttachment_2 Table',
        raw: '<viewdef>\n  <desc>The CollectionObjectAttachment Table</desc>\n  <definition>CollectionObjectAttachment_2</definition>\n</viewdef>',
        table: '[table CollectionObjectAttachment]',
        type: 'formtable',
      },
    ],
    views: [
      {
        altViews: {
          altViews: [
            {
              default: false,
              mode: 'view',
              name: 'CollectionObjectAttachment Icon View',
              viewDef: 'CollectionObjectAttachment IconView',
            },
            {
              default: false,
              mode: 'edit',
              name: 'CollectionObjectAttachment Icon Edit',
              viewDef: 'CollectionObjectAttachment IconView',
            },
            {
              default: false,
              mode: 'view',
              name: 'CollectionObjectAttachment Table View',
              viewDef: 'CollectionObjectAttachment Table',
            },
            {
              default: false,
              mode: 'edit',
              name: 'CollectionObjectAttachment Table Edit',
              viewDef: 'CollectionObjectAttachment Table',
            },
            {
              default: true,
              legacyLabel: 'Form',
              mode: 'view',
              name: 'CollectionObjectAttachment Form View',
              viewDef: 'CollectionObjectAttachment Form',
            },
            {
              default: false,
              legacyLabel: 'Form',
              mode: 'edit',
              name: 'CollectionObjectAttachment Form Edit',
              viewDef: 'CollectionObjectAttachment Form',
            },
          ],
        },
        description: 'The Collection Object-Attachment View.',
        businessRules:
          'edu.ku.brc.specify.datamodel.busrules.AttachmentBusRules',
        legacyResourceLabels: false,
        legacyUseBusinessRules: true,
        name: 'CollectionObjectAttachment',
        table: '[table CollectionObjectAttachment]',
        title: '',
      },
      {
        altViews: {
          altViews: [
            {
              default: false,
              legacyValidated: false,
              mode: 'view',
              name: 'CollectionObjectAttachment_2 View',
              viewDef: 'CollectionObjectAttachment_2',
            },
            {
              default: true,
              legacyValidated: false,
              mode: 'edit',
              name: 'CollectionObjectAttachment_2 Edit',
              viewDef: 'CollectionObjectAttachment_2',
            },
            {
              default: false,
              legacyValidated: false,
              mode: 'view',
              name: 'CollectionObjectAttachment_2 Table View',
              viewDef: 'CollectionObjectAttachment_2 Table',
            },
            {
              default: false,
              legacyValidated: false,
              mode: 'edit',
              name: 'CollectionObjectAttachment_2 Table Edit',
              viewDef: 'CollectionObjectAttachment_2 Table',
            },
            {
              default: false,
              legacyValidated: false,
              mode: 'view',
              name: 'CollectionObjectAttachment_2 Icon View',
              viewDef: 'CollectionObjectAttachment_2 Icon',
            },
            {
              default: false,
              legacyValidated: false,
              mode: 'edit',
              name: 'CollectionObjectAttachment_2 Icon Edit',
              viewDef: 'CollectionObjectAttachment_2 Icon',
            },
          ],
        },
        description: '',
        businessRules:
          'edu.ku.brc.specify.datamodel.busrules.AttachmentBusRules',
        legacyUseBusinessRules: true,
        name: 'CollectionObjectAttachment_2',
        table: '[table CollectionObjectAttachment]',
        title: 'CollectionObjectAttachment',
      },
    ],
    name: 'CollectionObjectAttachment',
  }));

test('Add new view definition based on existing', () =>
  expect(
    processViewSet(
      createViewDefinition(
        viewSets(),
        localized('A'),
        tables.Accession,
        viewDefinition
      )
    )
  ).toEqual({
    viewDefs: [
      {
        legacyEditableDialog: true,
        legacyGetTable: 'edu.ku.brc.af.ui.forms.DataGetterForObj',
        legacySetTable: 'edu.ku.brc.af.ui.forms.DataSetterForObj',
        name: 'CollectionObjectAttachment Table',
        raw: '<viewdef/>',
        table: '[table CollectionObjectAttachment]',
        type: 'formtable',
      },
      {
        legacyEditableDialog: true,
        legacyGetTable: 'edu.ku.brc.af.ui.forms.DataGetterForObj',
        legacySetTable: 'edu.ku.brc.af.ui.forms.DataSetterForObj',
        legacyUseResourceLabels: true,
        name: 'CollectionObjectAttachment Form',
        raw: '<viewdef/>',
        table: '[table CollectionObjectAttachment]',
        type: 'form',
      },
      {
        legacyEditableDialog: true,
        legacyGetTable: 'edu.ku.brc.af.ui.forms.DataGetterForObj',
        legacySetTable: 'edu.ku.brc.af.ui.forms.DataSetterForObj',
        name: 'CollectionObjectAttachment IconView',
        raw: '<viewdef/>',
        table: '[table CollectionObjectAttachment]',
        type: 'iconview',
      },
      {
        legacyEditableDialog: true,
        legacyGetTable: 'edu.ku.brc.af.ui.forms.DataGetterForObj',
        legacySetTable: 'edu.ku.brc.af.ui.forms.DataSetterForObj',
        legacyTable: 'edu.ku.brc.specify.datamodel.ObjectAttachmentIFace',
        legacyUseResourceLabels: true,
        name: 'ObjectAttachment Form',
        raw: '<viewdef/>',
        type: 'form',
      },
      {
        legacyEditableDialog: true,
        legacyGetTable: 'edu.ku.brc.af.ui.forms.DataGetterForObj',
        legacySetTable: 'edu.ku.brc.af.ui.forms.DataSetterForObj',
        name: 'CollectionObjectAttachment IconView (2)',
        raw: `<viewdef>
  <desc>The ObjectAttachment Icon Viewer</desc>
</viewdef>`,
        table: '[table CollectionObjectAttachment]',
        type: 'iconview',
      },
      {
        legacyEditableDialog: true,
        legacyGetTable: 'edu.ku.brc.af.ui.forms.DataGetterForObj',
        legacySetTable: 'edu.ku.brc.af.ui.forms.DataSetterForObj',
        name: 'CollectionObjectAttachment Table (2)',
        raw: `<viewdef>
  <desc>ObjectAttachment grid view.</desc>
  <definition>ObjectAttachment Form</definition>
</viewdef>`,
        table: '[table CollectionObjectAttachment]',
        type: 'formtable',
      },
      {
        legacyEditableDialog: true,
        legacyGetTable: 'edu.ku.brc.af.ui.forms.DataGetterForObj',
        legacySetTable: 'edu.ku.brc.af.ui.forms.DataSetterForObj',
        legacyUseResourceLabels: true,
        name: 'CollectionObjectAttachment Form (2)',
        raw: `<viewdef>
  <desc>The CollectionObjectAttachment form.</desc>
  <!-- <columnDef>110px,2dlu,p:g,5dlu,100px,2dlu,85px</columnDef> -->
  <columnDef>p,5dlu,p:g</columnDef>
  <rowDef auto='true' cell='p' sep='2px'/>
  <rows/>
</viewdef>`,
        table: '[table CollectionObjectAttachment]',
        type: 'form',
      },
    ],
    views: [
      {
        altViews: {
          altViews: [
            {
              default: false,
              mode: 'view',
              name: 'CollectionObjectAttachment Icon View',
              viewDef: 'CollectionObjectAttachment IconView',
            },
            {
              default: false,
              mode: 'edit',
              name: 'CollectionObjectAttachment Icon Edit',
              viewDef: 'CollectionObjectAttachment IconView',
            },
            {
              default: false,
              mode: 'view',
              name: 'CollectionObjectAttachment Table View',
              viewDef: 'CollectionObjectAttachment Table',
            },
            {
              default: false,
              mode: 'edit',
              name: 'CollectionObjectAttachment Table Edit',
              viewDef: 'CollectionObjectAttachment Table',
            },
            {
              default: true,
              legacyLabel: 'Form',
              mode: 'view',
              name: 'CollectionObjectAttachment Form View',
              viewDef: 'CollectionObjectAttachment Form',
            },
            {
              default: false,
              legacyLabel: 'Form',
              mode: 'edit',
              name: 'CollectionObjectAttachment Form Edit',
              viewDef: 'CollectionObjectAttachment Form',
            },
          ],
        },
        description: 'The Collection Object-Attachment View.',
        businessRules:
          'edu.ku.brc.specify.datamodel.busrules.AttachmentBusRules',
        legacyResourceLabels: false,
        legacyUseBusinessRules: true,
        name: 'CollectionObjectAttachment',
        table: '[table CollectionObjectAttachment]',
        title: '',
      },
      {
        altViews: {
          altViews: [
            {
              default: false,
              mode: 'view',
              name: 'CollectionObjectAttachment Icon View',
              viewDef: 'CollectionObjectAttachment IconView (2)',
            },
            {
              default: false,
              mode: 'edit',
              name: 'CollectionObjectAttachment Icon Edit',
              viewDef: 'CollectionObjectAttachment IconView (2)',
            },
            {
              default: false,
              mode: 'view',
              name: 'CollectionObjectAttachment Table View',
              viewDef: 'CollectionObjectAttachment Table (2)',
            },
            {
              default: false,
              mode: 'edit',
              name: 'CollectionObjectAttachment Table Edit',
              viewDef: 'CollectionObjectAttachment Table (2)',
            },
            {
              default: true,
              legacyLabel: 'Form',
              mode: 'view',
              name: 'CollectionObjectAttachment Form View',
              viewDef: 'CollectionObjectAttachment Form (2)',
            },
            {
              default: false,
              legacyLabel: 'Form',
              mode: 'edit',
              name: 'CollectionObjectAttachment Form Edit',
              viewDef: 'CollectionObjectAttachment Form (2)',
            },
          ],
        },
        description: 'The Collection Object-Attachment View.',
        businessRules:
          'edu.ku.brc.specify.datamodel.busrules.AttachmentBusRules',
        legacyResourceLabels: false,
        legacyUseBusinessRules: true,
        name: 'A',
        table: '[table CollectionObjectAttachment]',
      },
    ],
    name: 'CollectionObjectAttachment',
  }));
