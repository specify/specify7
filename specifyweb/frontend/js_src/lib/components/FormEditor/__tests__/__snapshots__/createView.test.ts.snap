// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tables with form tables computed correctly 1`] = `
[
  "[table Accession]",
  "[table AccessionAgent]",
  "[table AccessionAttachment]",
  "[table AccessionAuthorization]",
  "[table AccessionCitation]",
  "[table Address]",
  "[table AddressOfRecord]",
  "[table Agent]",
  "[table AgentAttachment]",
  "[table AgentGeography]",
  "[table AgentIdentifier]",
  "[table AgentSpecialty]",
  "[table Appraisal]",
  "[table Attachment]",
  "[table Author]",
  "[table Borrow]",
  "[table BorrowAgent]",
  "[table BorrowAttachment]",
  "[table BorrowMaterial]",
  "[table BorrowReturnMaterial]",
  "[table CollectingEvent]",
  "[table CollectingEventAttachment]",
  "[table CollectingEventAttr]",
  "[table CollectingEventAttribute]",
  "[table CollectingEventAuthorization]",
  "[table CollectingTrip]",
  "[table CollectingTripAttachment]",
  "[table CollectingTripAttribute]",
  "[table CollectingTripAuthorization]",
  "[table Collection]",
  "[table CollectionObjectAttachment]",
  "[table CollectionObjectAttr]",
  "[table CollectionObjectAttribute]",
  "[table CollectionObjectCitation]",
  "[table CollectionObjectProperty]",
  "[table CollectionRelType]",
  "[table CollectionRelationship]",
  "[table Collector]",
  "[table CommonNameTx]",
  "[table CommonNameTxCitation]",
  "[table ConservDescription]",
  "[table ConservDescriptionAttachment]",
  "[table ConservEvent]",
  "[table ConservEventAttachment]",
  "[table Container]",
  "[table DNAPrimer]",
  "[table DNASequence]",
  "[table DNASequenceAttachment]",
  "[table DNASequencingRun]",
  "[table DNASequencingRunAttachment]",
  "[table DNASequencingRunCitation]",
  "[table Deaccession]",
  "[table DeaccessionAgent]",
  "[table DeaccessionAttachment]",
  "[table Determination]",
  "[table DeterminationCitation]",
  "[table Determiner]",
  "[table Discipline]",
  "[table Disposal]",
  "[table DisposalAgent]",
  "[table DisposalAttachment]",
  "[table DisposalPreparation]",
  "[table Division]",
  "[table ExchangeIn]",
  "[table ExchangeInAttachment]",
  "[table ExchangeInPrep]",
  "[table ExchangeOut]",
  "[table ExchangeOutAttachment]",
  "[table ExchangeOutPrep]",
  "[table Exsiccata]",
  "[table ExsiccataItem]",
  "[table Extractor]",
  "[table FieldNotebook]",
  "[table FieldNotebookAttachment]",
  "[table FieldNotebookPage]",
  "[table FieldNotebookPageAttachment]",
  "[table FieldNotebookPageSet]",
  "[table FieldNotebookPageSetAttachment]",
  "[table FundingAgent]",
  "[table GeoCoordDetail]",
  "[table Geography]",
  "[table GeographyTreeDef]",
  "[table GeographyTreeDefItem]",
  "[table GeologicTimePeriodTreeDef]",
  "[table GeologicTimePeriodTreeDefItem]",
  "[table Gift]",
  "[table GiftAgent]",
  "[table GiftAttachment]",
  "[table GiftPreparation]",
  "[table GroupPerson]",
  "[table InfoRequest]",
  "[table Institution]",
  "[table InstitutionNetwork]",
  "[table Journal]",
  "[table LatLonPolygon]",
  "[table LatLonPolygonPnt]",
  "[table LithoStratTreeDef]",
  "[table LithoStratTreeDefItem]",
  "[table Loan]",
  "[table LoanAgent]",
  "[table LoanAttachment]",
  "[table LoanPreparation]",
  "[table LoanReturnPreparation]",
  "[table Locality]",
  "[table LocalityAttachment]",
  "[table LocalityCitation]",
  "[table LocalityDetail]",
  "[table LocalityNameAlias]",
  "[table MaterialSample]",
  "[table OtherIdentifier]",
  "[table PcrPerson]",
  "[table Permit]",
  "[table PermitAttachment]",
  "[table PrepType]",
  "[table Preparation]",
  "[table PreparationAttachment]",
  "[table PreparationAttr]",
  "[table PreparationAttribute]",
  "[table PreparationProperty]",
  "[table Project]",
  "[table ReferenceWork]",
  "[table ReferenceWorkAttachment]",
  "[table RepositoryAgreement]",
  "[table RepositoryAgreementAttachment]",
  "[table Shipment]",
  "[table Storage]",
  "[table StorageAttachment]",
  "[table StorageTreeDef]",
  "[table StorageTreeDefItem]",
  "[table Taxon]",
  "[table TaxonAttachment]",
  "[table TaxonAttribute]",
  "[table TaxonCitation]",
  "[table TaxonTreeDef]",
  "[table TaxonTreeDefItem]",
  "[table TreatmentEvent]",
  "[table TreatmentEventAttachment]",
  "[table VoucherRelationship]",
]
`;
